manpages/findhotpixels.man
.na
.nh
.TH findhotpixels 1 4.6.34 BL3DEMC
.SH NAME
findhotpixels - find pixels to exclude from X-ray analysis in SerialEM
.SH SYNOPSIS
findhotpixels  [options]  input_image  output_model
.SH DESCRIPTION
Findhotpixels will analyze a set of CCD camera dark reference images to
identify relatively hot pixels in the CCD chip and will output a list of the
pixels in a format suitable for placing in the property file for SerialEM.
Automatic removal of X-rays from dark references can then be activated in
SerialEM, and it will then be able to avoid identifying a hot pixel as an
X-ray.  It is recommended that three long-exposure (e.g., 3 sec) dark
reference images be used for this analysis.
.P
The program works by finding pixels that deviation from the image mean by a
specified threshold.  The lower the threshold, the more pixels it will find.
You may need to run the program several times to find a threshold that gives
a reasonable number of hot pixels (under 100).  A threshold on the order of
0.2% of the dynamic range of the camera should be appropriate.
.P
Hot columns are excluded from the analysis.  With 4-port readout cameras
from Gatan, it is important to exclude a fairly large number of bright
columns near the edges; otherwise you will need to set the threshold higher
than the optimal level for finding the hot pixels in the rest of the
camera.  The program could be smarter about this and use a deviation from a
local mean, but there is no point in doing so under SerialEM uses a local
criterion as well.
.P
The program outputs a list of the hot pixels coordinates and their values
above the image mean.  At the end is a set of HotPixels and HotColumns
entries suitable for pasting into the SerialEM properties file.
.SH OPTIONS
The program can be run either interactively or with command line arguments
processed by the PIP interface (see the man page for pip(1)).  With command
line arguments, the options are:
.TP
.B -input \fROR\fP -InputFile 	 \fIFile name\fR
Name of input image file with dark reference images.  If this option is
not entered, the first non-option argument is used
.TP
.B -output \fROR\fP -OutputFile 	 \fIFile name\fR
Name of output model file with a point at every hot pixel.  If this option
is not entered, the second non-option argument is used.
.TP
.B -columns \fROR\fP -HotColumns 	 \fIList of integer ranges\fR
List of hot columns to exclude (numbered from 0).  Ranges may be entered.
.TP
.B -threshold \fROR\fP -ThresholdDifference 	 \fIFloating point\fR
Threshold difference from mean value for seeking hot pixels
.TP
.B -help \fROR\fP -usage
Print help output
.TP
.B -StandardInput
Read parameter entries from standard input.
.P
.SH INTERACTIVE INPUTS
If no command line arguments are entered, input is requested interactively
as follows:
.P
Name of input image file with dark reference images
.P
Name of output model file with a point at every hot pixel
.P
List of hot columns to exclude (numbered from 0)
.P
Threshold difference from mean value for seeking hot pixels
.P
.SH HISTORY
Written by David Mastronarde in 2003.
Added to IMOD, 10/17/06.
.SH BUGS
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names
will differ in different Java versions.  This bug will not be fixed because it
does not appear to cause any harm.  Only the automated tests where modified.
(Bug# 1753)


flib/model/findhotpixels.f
c       FINDHOTPIXELS
c       
c         Findhotpixels will analyze a set of CCD camera dark reference images
c       to identify relatively hot pixels in the CCD chip and will output a
c       list of the pixels in a format suitable for placing in the property
c       file for SerialEM.  See man page and SerialEM help manual for more
c       details.
c       
c       David Mastronarde, 2003.
c       
c       $Id$
c       
c       $Log$
c       Revision 3.2  2006/10/18 14:41:05  mast
c       Fixed option line for all those other compilers
c
c       Revision 3.1  2006/10/17 22:57:31  mast
c       Added to package
c
c
      implicit none
      integer idim,limsec
      parameter (idim=20480,limsec=10)
      integer*4 NX,NY,NZ,NXYZ(3),MXYZ(3),NXYZST(3)
      COMMON //NX,NY,NZ
      include 'smallmodel.inc'
C       
      real*4 ARRAY(idim,limsec),TITLE(20)
C       
      CHARACTER*120 FILIN,FILOUT
      character*512 colstring
C       
      integer*4 listcols(5000)
      EQUIVALENCE (NX,NXYZ)
      integer*4 nhotcols, ierr, nzdo, mode, ix, iy, iz, ifonlist, j, nsum, i
      real*4 thresh, dmin,dmax,dmean, sum, avg
c       
      logical pipinput
      integer*4 numOptArg, numNonOptArg
      integer*4 PipGetString,PipGetFloat
      integer*4 PipGetInOutFile, PipParseInput, PipGetBoolean
      integer numOptions
      parameter (numOptions = 5)
      character*(120 * numOptions) options(1) 
      options(1) = 
     &    'input:InputFile:FN:Name of input image file@'//
     &    'output:OutputFile:FN:Name of output model file@'//
     &    'columns:HotColumns:LI:List of hot columns to exclude '//
     &    '(numbered from 0)@threshold:ThresholdDifference:F:'//
     &    'Threshold difference from mean value for hot pixels@'//
     &    'help:usage:B:Print help output'
c       
      nhotcols = 0
      call PipExitOnError(0, 'ERROR: FINDHOTPIXELS - ')
      call PipAllowCommaDefaults(1)
      ierr = PipParseInput(options, numOptions, '@', numOptArg, numNonOptArg)
      pipinput = numOptArg + numNonOptArg .gt. 0
      if ((pipinput .and. numOptArg + numNonOptArg .lt. 3)  .or.
     &    PipGetBoolean('help', j) .eq. 0) then
        call PipPrintHelp('findhotpixels', 0, 1, 1)
        call exit(0)
      endif
c       
      if (PipGetInOutFile('InputFile', 1, 'Name of input image file', filin)
     &    .ne. 0) call exitError('NO INPUT FILE SPECIFIED')
      if (PipGetInOutFile('OutputFile', 2, 'Name of output model file', filout)
     &    .ne. 0) call exitError('NO OUTPUT FILE SPECIFIED')
      if (pipinput) then
        if (PipGetString('HotColumns', colstring) .eq. 0)
     &      call parselist(colstring, listcols, nhotcols)
        if (PipGetFloat('ThresholdDifference', thresh) .ne. 0)
     &      call exitError('YOU MUST ENTER A THRESHOLD DIFFERENCE FROM MEAN')
      else
        print *,'Enter list of hot columns to exclude (numbered from',
     &      ' 0), or Return if none'
        call rdlist(5, listcols, nhotcols)
c         
        write(*,'(1x,a,$)')'Threshold difference from mean: '
        read(5,*)thresh
      endif
C       
      call imopen(1,filin,'ro')
      call irdhdr(1,nxyz,mxyz,mode,dmin,dmax,dmean)
      nzdo = min(nz,limsec)

      print *
      thresh = thresh + dmean
      max_mod_obj = 1
      ibase_obj(1) = 0
      obj_color(1,1)=1
      obj_color(2,1)=255
      n_point = 0
      do iy = 1, ny
        do iz=1,nzdo
          call imposn(1, iz - 1, iy - 1)
          call irdlin(1, array(1,iz), *99)
        enddo
        
        do ix = 1, nx

          ifonlist = 0
          do j = 1, nhotcols
            if (ix-1 .eq. listcols(j)) ifonlist = 1
          enddo

          if (ifonlist .eq. 0) then
            sum = 0.
            nsum = 0
            do iz = 1, nzdo
              if (array(ix,iz).gt.thresh)then
                nsum=nsum+1
                sum=sum+array(ix,iz)-dmean
              endif
            enddo
            if (nsum.eq.nzdo)then
              avg=sum/nsum  
              n_point=n_point+1
              if (n_point .gt. max_pt) call exitError(
     &            'TOO MANY POINTS ABOVE THRESHOLD; RAISE THE THRESHOLD')
              p_coord(1,n_point) = ix - 0.55
              p_coord(2,n_point) = iy - 0.55
              p_coord(3,n_point) = nzdo / 2
              object(n_point)=n_point
              write(*,'(2i6,f7.0)') ix,iy,avg
            endif
          endif
        enddo
      enddo

      npt_in_obj(1) = n_point
      call putimodflag(1, 2)
      call putsymtype(1, 0)
      call putsymsize(1, 5)
      call write_wmod(filout)

      print *
      PRINT *,'Total pixels: ',n_point
      print *
      do i = 1,n_point
        if (mod(i,5) .eq. 1) write(*,'(a,$)') 'HotPixels '
        write(*,'(2i6,$)') nint(p_coord(1,i) + 0.5),nint(p_coord(2,i) + 0.5)
        if (mod(i,5).eq.0) print *
      enddo
      print *
      print *
      do i = 1, nhotcols
        if (mod(i,13) .eq. 1) write(*,'(a,$)') 'HotColumns '
        write(*,'(i5,$)') listcols(i)
        if (mod(i,13).eq.0) print *
      enddo
      print *
c
      call imclose(1)
      call exit(0)
99    call exitError('READING FILE')
      end


